File: frontend/src/lib/mappers/messageMapper.ts
import { ReadMessageDto } from "@/lib/dto/message/ReadMessageDto";
import { Message, MessageStatus } from "@/lib/types";
import { mapMessageAttachmentDtos } from "./attachmentMapper";

export function mapReadMessageDto(dto: ReadMessageDto): Message {
  return {
    id: dto.id,
    text: dto.text,
    userId: dto.userId,
    chatId: dto.chatId,
    status: (typeof dto.status === 'number' ? dto.status : dto.status) as MessageStatus,
    createdAt: new Date(dto.createdAt),
    attachments: mapMessageAttachmentDtos(dto.attachments || null)
  };
}

export function mapReadMessageDtos(dtos: ReadMessageDto[]): Message[] {
  return dtos.map(mapReadMessageDto);
}

File: user-service/Application/Mappings/MessageMapper.cs
using Application.DTOs.Message;
using Application.Events;
using Domain.Entities;

namespace Application.Mappings;

/// <summary>
/// Mapper for message entities to DTOs and events
/// </summary>
public static class MessageMapper
{
    public static MessageCreatedEvent MapToEvent(
        MessageEntity message, 
        IEnumerable<MessageAttachmentEntity> attachments)
    {
        return new MessageCreatedEvent
        {
            MessageId = message.Id,
            ChatId = message.ChatId,
            Text = message.Text,
            UserId = message.UserId,
            CreatedAt = message.CreatedAt,
            Status = (int)message.Status,
            Attachments = attachments.Select(a => new MessageAttachmentEventDto
            {
                Id = a.Id,
                Name = a.FileName,
                ContentType = a.MimeType,
                Size = a.FileSize,
                Url = $"/api/Attachment/download/{a.Id}"
            }).ToList()
        };
    }

    public static ReadMessageDto MapToDto(
        MessageEntity message, 
        IEnumerable<MessageAttachmentEntity> attachments)
    {
        return new ReadMessageDto
        {
            Id = message.Id,
            ChatId = message.ChatId,
            Text = message.Text,
            UserId = message.UserId,
            CreatedAt = message.CreatedAt,
            Status = message.Status,
            Attachments = attachments.Select(a => new MessageAttachmentDto
            {
                Id = a.Id,
                Name = a.FileName,
                ContentType = a.MimeType,
                Size = a.FileSize,
                Url = $"/api/Attachment/download/{a.Id}"
            }).ToList()
        };
    }
}
